web/src/engine/predictive-text/worker-thread/src/main/models/composited-traversal.ts

```diff
@@ -0,0 +1,256 @@
+import { PriorityQueue } from "@keymanapp/web-utils";
+
+export class CompositedTraversal implements LexiconTraversal {
+  private readonly componentTraversals: Distribution<LexiconTraversal>;
+
+  // Used both internally and externally - and we don't want to re-weight internal calls!
+  constructor(traversalsWithWeighting: Distribution<LexiconTraversal>) {
+    // let totalP = traversalsWithWeighting.reduce((accum, current) => accum + current.p, 0);
+    this.componentTraversals = traversalsWithWeighting.map((entry) => {
+      // Shallow-copy the traversals while normalizing the associated weightings.
+      // The sum of all `p` values should sum to 1.
+      return {
+        sample: entry.sample,
+        p: entry.p // totalP
+      };
+    });
+  }
+
+  children(): { char: string; p: number, traversal: () => LexiconTraversal }[] {
```

Comment:
Can we make `{ char: string; p: number, traversal: () => LexiconTraversal }` into a local type as it is used in a couple of places